functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+   ([db ent-id attr-name ts] (:value (attr-at db ent-id attr-name ts))))
+
+(defn indx-at
+   ([db kind] (indx-at db kind (:curr-time db)))
+   ([db kind ts] (kind ((:layers db) ts))))
+````
+
+Since we treat our database just like any other value, each of these functions take a database as an argument. Each element is retrieved by its associated identifier, and optionally the timestamp of interest. This timestamp is used in each case to find the corresponding layer that our lookup should be applied to.
+
+#### Evolution
+
+A first usage of the basic accessors is to provide a "read-into-the-past" API. This is possible as in our database, an update operation is done by appending a new layer (as oppose to overwriting). Therefore we can use the *prev-ts* property of an attribute look at the attribute at that layer, and continue going back in time looking deeper into history, thus observing how the attribute’s value evolved throughout time.  
+
+The function *evolution-of* does exactly that, and return a sequence of pairs - each consist of the timestamp and value of an attribute’s update.
+````clojure
+(defn evolution-of [db ent-id attr-name]
+   (loop [res [] ts (:curr-time db)]
+     (if (= -1 ts) (reverse res)
+         (let [attr (attr-at db ent-id attr-name ts)]
+           (recur (conj res {(:ts attr) (:value attr)})  (:prev-ts attr))))))
+````
+## Data behavior and life cycle
+
+So far, our discussion has focused on the structure of our data -- what the core components are, and how they are aggregated together. It's time now to explore the dynamics of our system; how data is changed over time through the _data lifecycle_ (add => update => remove). 
+
+As we've already discussed, data in an archaeologist's world never actually changes. Once it is created, it exists forever and can only be hidden from the world by data in a newer layer. Note the term ‘hidden’, as it is crucial here. Older data does not 'disappear' -- it is buried, and can be revealed again by exposing an older layer. Conversely, updating data means obscuring the old by adding a new layer on top of it with something else. We can thus 'delete' data by adding a layer of 'nothing' on top of it. 
+
+This means that when talking about data lifecycle, we are really talking about adding layers to our data over time. 
+
+### The bare necessities
+
+The data lifecycle consists of three basic operations, which we will discuss here:
+
+* adding an entity with the *add-entity* function
+* removing an entity with the *remove-entity* function
+* updating an entity with the *update-entity* function. 
+
+Remember that, even though these functions provide the illusion of mutability, all that we are really doing in each case is adding another layer to the data. Also, since we use here Clojure's immutable data structures, we pay for such operations the price of an "in-place" change from the caller's perspective, while maintaining immutability for all other users of that data structure.

```

Comment:
change immutable to persistent data structures, this is a huge topic, so I prefer not to enter it (just say that it is the enabler for the db capabilities)